README.md
# intl-data

Tiny module to retrieve locale and currency data from native builtin Intl object without locale data transfer or additional locale files.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Installation](#installation)
- [Synopsis](#synopsis)
- [Details](#details)
- [API](#api)
- [intl-data](#intl-data)
  - [Table of contents](#table-of-contents)
    - [Interfaces](#interfaces)
    - [Functions](#functions)
  - [Functions](#functions-1)
    - [getCurrency](#getcurrency)
    - [getLocale](#getlocale)
- [Interfaces](#interfaces-1)
- [Interface: Currency](#interface-currency)
  - [Hierarchy](#hierarchy)
  - [Table of contents](#table-of-contents-1)
    - [Properties](#properties)
  - [Properties](#properties-1)
    - [code](#code)
    - [decimalPlaces](#decimalplaces)
    - [name](#name)
    - [symbol](#symbol)
- [Interface: Locale](#interface-locale)
  - [Hierarchy](#hierarchy-1)
  - [Table of contents](#table-of-contents-2)
    - [Properties](#properties-2)
  - [Properties](#properties-3)
    - [code](#code-1)
    - [currencySymbolPlacement](#currencysymbolplacement)
    - [decimalCharacter](#decimalcharacter)
    - [digitGroupSeparator](#digitgroupseparator)
    - [percentSymbol](#percentsymbol)
    - [percentSymbolPlacement](#percentsymbolplacement)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Installation

| Format                             | Instructions                                           |
| ---------------------------------- | ------------------------------------------------------ |
| **TypeScript**                     | `import { ... } from "intl-data";`                     |
| **Node (CommonJS)**                | `const { ... } = require("intl-data");`                |
| **ES Modules (Browsers, webpack)** | `import { ... } from "intl-data";`                     |
| **UMD (Browser Global)**           | `<script src="path/to/umd/intl-data.js"></script>`     |
| **UMD (Browser Global)** (min)     | `<script src="path/to/umd/intl-data.min.js"></script>` |

All unminified and minified bundle files are located in respective directories named after format (cjs, esm, umd) in `dist` (`dist/cjs/intl-data.js` and `intl-data.min.js`)

# Synopsis

```ts
import { getLocale, getCurrency } from "intl-data";

const locale = getLocale("tr-TR"); // { digitGroupSeparator: ".", decimalCharacter: ",", currencySymbolPlacement: "p", percentSymbolPlacement: "p", percentSymbol: "%" }
const currency = getCurrency("TRY"); // { symbol: "₺", decimalPlaces: 2 }
const currency = getCurrency("TRY", "en-US"); // { symbol: '₺', decimalPlaces: 2, name: 'Turkish Lira' }
```

# Details

Retrieves locale and currency.

Modern browsers and node.js have builtin [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) object, and this module uses this object. As a result it does not need to transfer additional locale files or data.

# API

<a name="readmemd"></a>

intl-data

# intl-data

## Table of contents

### Interfaces

- [Currency](#interfacescurrencymd)
- [Locale](#interfaceslocalemd)

### Functions

- [getCurrency](#getcurrency)
- [getLocale](#getlocale)

## Functions

### getCurrency

▸ **getCurrency**<T\>(`currencyCode`: T, `locale?`: _string_): T

#### Type parameters:

| Name | Type        |
| ---- | ----------- | ------ |
| `T`  | _undefined_ | _null_ |

#### Parameters:

| Name           | Type     |
| -------------- | -------- |
| `currencyCode` | T        |
| `locale?`      | _string_ |

**Returns:** T

Defined in: [intl-data.ts:37](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L37)

▸ **getCurrency**(`currencyCode`: _string_ | _null_ | _undefined_, `locale?`: _string_): [_Currency_](#interfacescurrencymd) | _null_ | _undefined_

#### Parameters:

| Name           | Type     |
| -------------- | -------- | ------ | ----------- |
| `currencyCode` | _string_ | _null_ | _undefined_ |
| `locale?`      | _string_ |

**Returns:** [_Currency_](#interfacescurrencymd) | _null_ | _undefined_

Defined in: [intl-data.ts:38](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L38)

---

### getLocale

▸ **getLocale**(`localeCode`: _string_): [_Locale_](#interfaceslocalemd)

#### Parameters:

| Name         | Type     |
| ------------ | -------- |
| `localeCode` | _string_ |

**Returns:** [_Locale_](#interfaceslocalemd)

Defined in: [intl-data.ts:79](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L79)

▸ **getLocale**<T\>(`localeCode`: T): T

#### Type parameters:

| Name | Type        |
| ---- | ----------- | ------ |
| `T`  | _undefined_ | _null_ |

#### Parameters:

| Name         | Type |
| ------------ | ---- |
| `localeCode` | T    |

**Returns:** T

Defined in: [intl-data.ts:80](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L80)

▸ **getLocale**(`localeCode`: _string_ | _null_ | _undefined_): [_Locale_](#interfaceslocalemd) | _undefined_ | _null_

#### Parameters:

| Name         | Type     |
| ------------ | -------- | ------ | ----------- |
| `localeCode` | _string_ | _null_ | _undefined_ |

**Returns:** [_Locale_](#interfaceslocalemd) | _undefined_ | _null_

Defined in: [intl-data.ts:81](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L81)

# Interfaces

<a name="interfacescurrencymd"></a>

[intl-data](#readmemd) / Currency

# Interface: Currency

Information about currency

## Hierarchy

- **Currency**

## Table of contents

### Properties

- [code](#code)
- [decimalPlaces](#decimalplaces)
- [name](#name)
- [symbol](#symbol)

## Properties

### code

• **code**: _string_

ISO 4217 currency codes

Defined in: [intl-data.ts:27](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L27)

---

### decimalPlaces

• **decimalPlaces**: _number_

Number of digits used in decimal (fractional) part of the currency.

Defined in: [intl-data.ts:29](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L29)

---

### name

• **name**: _string_

Name of the currency.

Defined in: [intl-data.ts:25](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L25)

---

### symbol

• **symbol**: _string_

Currency symbol such as `₺`, `$`, `€`

Defined in: [intl-data.ts:31](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L31)

<a name="interfaceslocalemd"></a>

[intl-data](#readmemd) / Locale

# Interface: Locale

Information about locale.

## Hierarchy

- **Locale**

## Table of contents

### Properties

- [code](#code)
- [currencySymbolPlacement](#currencysymbolplacement)
- [decimalCharacter](#decimalcharacter)
- [digitGroupSeparator](#digitgroupseparator)
- [percentSymbol](#percentsymbol)
- [percentSymbolPlacement](#percentsymbolplacement)

## Properties

### code

• **code**: _string_

BCP 47 language tag of the locale. This is simply string provided to get locale. Example: `tr-TR`

Defined in: [intl-data.ts:9](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L9)

---

### currencySymbolPlacement

• **currencySymbolPlacement**: _p_ | _s_

Place of currency symbol: `p` for prefix such as `$1.95`, or `s` for suffix such as `1.95$`

Defined in: [intl-data.ts:15](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L15)

---

### decimalCharacter

• **decimalCharacter**: _string_

Character to separate decimal (fraction) parts of the number from integer parts.

Defined in: [intl-data.ts:11](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L11)

---

### digitGroupSeparator

• **digitGroupSeparator**: _string_

Grouping separators for integer parts, such as thousands separators or thousand/lakh/crore separators.

Defined in: [intl-data.ts:13](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L13)

---

### percentSymbol

• **percentSymbol**: _string_

Percent symbol. Some locales contains spcae for percent symbol such as `12 %`, and some not `12%`. This is used to differenciate them.

Defined in: [intl-data.ts:19](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L19)

---

### percentSymbolPlacement

• **percentSymbolPlacement**: _p_ | _s_

Place of percent symbol: `p` for prefix such as `%12`, or `s` for suffix such as `12%`

Defined in: [intl-data.ts:17](https://github.com/ozum/intl-data/blob/ac8c5c1/src/intl-data.ts#L17)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.0.10 (2021-01-25)

### Bug Fixes

- fix currency symbol for different locales with same currency ([e15ad83](https://github.com/ozum/intl-data/commit/e15ad83b7b55ba9bc442d623ae741a168a526217)), closes [#1](https://github.com/ozum/intl-data/issues/1)
- fix es-CO locale ([de72d13](https://github.com/ozum/intl-data/commit/de72d131073171abf4472a2cd68e5bdb7a62f153))
- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix type definitions ([e90e92a](https://github.com/ozum/intl-data/commit/e90e92a6d8d60a3ba244e414592d1f4103959554))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))
- update dev packages ([4bc9bb2](https://github.com/ozum/intl-data/commit/4bc9bb28a35e689619fedfc95b43809786f4aa4c))

### 1.0.9 (2020-05-06)

### Bug Fixes

- fix currency symbol for different locales with same currency ([e15ad83](https://github.com/ozum/intl-data/commit/e15ad83b7b55ba9bc442d623ae741a168a526217)), closes [#1](https://github.com/ozum/intl-data/issues/1)
- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix type definitions ([e90e92a](https://github.com/ozum/intl-data/commit/e90e92a6d8d60a3ba244e414592d1f4103959554))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))
- update dev packages ([4bc9bb2](https://github.com/ozum/intl-data/commit/4bc9bb28a35e689619fedfc95b43809786f4aa4c))

### 1.0.8 (2020-03-25)

### Bug Fixes

- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix type definitions ([e90e92a](https://github.com/ozum/intl-data/commit/e90e92a6d8d60a3ba244e414592d1f4103959554))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))
- update dev packages ([4bc9bb2](https://github.com/ozum/intl-data/commit/4bc9bb28a35e689619fedfc95b43809786f4aa4c))

### 1.0.7 (2020-03-19)

### Bug Fixes

- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix type definitions ([e90e92a](https://github.com/ozum/intl-data/commit/e90e92a6d8d60a3ba244e414592d1f4103959554))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))
- update dev packages ([4bc9bb2](https://github.com/ozum/intl-data/commit/4bc9bb28a35e689619fedfc95b43809786f4aa4c))

### 1.0.6 (2020-03-18)

### Bug Fixes

- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))
- update dev packages ([4bc9bb2](https://github.com/ozum/intl-data/commit/4bc9bb28a35e689619fedfc95b43809786f4aa4c))

### 1.0.5 (2020-03-17)

### Bug Fixes

- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- fix typescript types ([0f10566](https://github.com/ozum/intl-data/commit/0f105667569aa6bfda714a2da9d4b5ae6f93bb17))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))
- update dependencies ([bd46577](https://github.com/ozum/intl-data/commit/bd46577c7b5af0df000229b524d226397c42f8ee))

### 1.0.4 (2019-11-13)

### Bug Fixes

- fix getLocale types ([d54f1b3](https://github.com/ozum/intl-data/commit/d54f1b319d5469010afd7018450566c7037f759a))
- fix homepage in package.json ([661791f](https://github.com/ozum/intl-data/commit/661791f66d3d3f0c57be85b8ce2f21591bb4e7a5))
- switch to rollup ([70707b4](https://github.com/ozum/intl-data/commit/70707b4f72f07e3dc90c809c26483cdb56b92764))

src/intl-data.test.ts
import { getLocale, getCurrency, Locale } from "./intl-data";

const trLocale = {
  code: "tr-TR",
  digitGroupSeparator: ".",
  decimalCharacter: ",",
  currencySymbolPlacement: "p",
  percentSymbolPlacement: "p",
  percentSymbol: "%",
};

/**
 * Checks whether given locale produces expected values by ignoring specila accented characters.
 *
 * @param locale is the locale to test.
 * @param expected is the expected values as an array of "code", "currencySymbolPlacement", "decimalCharacter", "digitGroupSeparator", "percentSymbol", "percentSymbolPlacement".
 * @returns whether given locale's result are equal to expected values.
 */
function isEqual(locale: string, expected: string[]): boolean {
  const result = getLocale(locale) as Record<string, any>;
  return Object.keys(result)
    .sort()
    .every((key, i) => result[key].localeCompare(expected[i], "en", { sensitivity: "base" }) === 0);
}

describe("getLocale", () => {
  it("should return locale for given locale code with prefix.", () => {
    expect(isEqual("tr-TR", ["tr-TR", "p", ",", ".", "%", "p"])).toBe(true);
  });

  it("should return locale for given locale code with suffix.", () => {
    expect(isEqual("de-DE", ["de-DE", "s", ",", ".", " %", "s"])).toBe(true);
  });

  it("should return locale from cache.", () => {
    expect(isEqual("tr-TR", ["tr-TR", "p", ",", ".", "%", "p"])).toBe(true);
  });

  it("should return null for null.", () => {
    expect(getLocale(null)).toBeNull();
  });

  it("should return undefined for undefined.", () => {
    expect(getLocale(undefined)).toBeUndefined();
  });

  it("should throw for unsupported locale.", () => {
    expect(() => getLocale("XYZ")).toThrow("Unsupported locale");
  });

  it("should return correct locale for es-CO", () => {
    expect(isEqual("es-CO", ["es-CO", "p", ",", ".", " %", "s"])).toBe(true);
  });
});

describe("getCurrency", () => {
  it("should return currency.", () => {
    expect(getCurrency("TRY", "tr-TR")).toEqual({ code: "TRY", symbol: "₺", decimalPlaces: 2, name: "Türk lirası" });
  });

  it("should return currency from cache.", () => {
    expect(getCurrency("TRY", "tr-TR")).toEqual({ code: "TRY", symbol: "₺", decimalPlaces: 2, name: "Türk lirası" });
  });

  it("should return currency without a symbol.", () => {
    expect(getCurrency("OMR")).toEqual({ code: "OMR", symbol: "OMR", decimalPlaces: 3, name: "Omani rials" });
  });

  it("should return null for null.", () => {
    expect(getCurrency(null)).toBeNull();
  });

  it("should return undefined for undefined.", () => {
    expect(getCurrency(undefined)).toBeUndefined();
  });

  it("should throw for unsupported locale.", () => {
    expect(() => getCurrency("XYZ")).toThrow("Invalid currency code");
  });

  it("should respect locale specific sign of the same currency", () => {
    const expected = {
      CNY: { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "Chinese yuan" },
      "CNY zh-CN": { code: "CNY", symbol: "¥", decimalPlaces: 2, name: "人民币" },
      "CNY zh-TW": { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "人民幣" },
      "CNY tr-TR": { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "Çin yuanı" },
    };

    const result = {
      CNY: getCurrency("CNY"),
      "CNY zh-CN": getCurrency("CNY", "zh-CN"),
      "CNY zh-TW": getCurrency("CNY", "zh-TW"),
      "CNY tr-TR": getCurrency("CNY", "tr-TR"),
    };

    expect(result).toEqual(expected);
  });
});


test/intl-data.test.ts
import { getLocale, getCurrency } from "../src/intl-data";

/**
 * Checks whether given locale produces expected values by ignoring specila accented characters.
 *
 * @param locale is the locale to test.
 * @param expected is the expected values as an array of "code", "currencySymbolPlacement", "decimalCharacter", "digitGroupSeparator", "percentSymbol", "percentSymbolPlacement".
 * @returns whether given locale's result are equal to expected values.
 */
function isEqual(locale: string, expected: string[]): boolean {
  const result = getLocale(locale) as Record<string, any>;
  return Object.keys(result)
    .sort()
    .every((key, i) => result[key].localeCompare(expected[i], "en", { sensitivity: "base" }) === 0);
}

describe("getLocale", () => {
  it("should return locale for given locale code with prefix.", () => {
    expect(isEqual("tr-TR", ["tr-TR", "p", ",", ".", "%", "p"])).toBe(true);
  });

  it("should return locale for given locale code with suffix.", () => {
    expect(isEqual("de-DE", ["de-DE", "s", ",", ".", " %", "s"])).toBe(true);
  });

  it("should return locale from cache.", () => {
    expect(isEqual("tr-TR", ["tr-TR", "p", ",", ".", "%", "p"])).toBe(true);
  });

  it("should return null for null.", () => {
    expect(getLocale(null)).toBeNull();
  });

  it("should return undefined for undefined.", () => {
    expect(getLocale(undefined)).toBeUndefined();
  });

  it("should throw for unsupported locale.", () => {
    expect(() => getLocale("XYZ")).toThrow("Unsupported locale");
  });

  it("should return correct locale for es-CO", () => {
    expect(isEqual("es-CO", ["es-CO", "p", ",", ".", " %", "s"])).toBe(true);
  });

  it("should return correct locale for ja-JP", () => {
    expect(isEqual("ja-JP", ["ja-JP", "p", ".", ",", "%", "s"])).toBe(true);
  });
});

describe("getCurrency", () => {
  it("should return currency.", () => {
    expect(getCurrency("TRY", "tr-TR")).toEqual({ code: "TRY", symbol: "₺", decimalPlaces: 2, name: "Türk lirası" });
  });

  it("should return currency from cache.", () => {
    expect(getCurrency("TRY", "tr-TR")).toEqual({ code: "TRY", symbol: "₺", decimalPlaces: 2, name: "Türk lirası" });
  });

  it("should return currency without a symbol.", () => {
    expect(getCurrency("OMR", "en-US")).toEqual({ code: "OMR", symbol: "OMR", decimalPlaces: 3, name: "Omani rials" });
  });

  it("should return null for null.", () => {
    expect(getCurrency(null)).toBeNull();
  });

  it("should return undefined for undefined.", () => {
    expect(getCurrency(undefined)).toBeUndefined();
  });

  it("should throw for unsupported locale.", () => {
    expect(() => getCurrency("XYZ")).toThrow("Invalid currency code");
  });

  it("should respect locale specific sign of the same currency", () => {
    const expected = {
      CNY: { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "Chinese yuan" },
      "CNY zh-CN": { code: "CNY", symbol: "¥", decimalPlaces: 2, name: "人民币" },
      "CNY zh-TW": { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "人民幣" },
      "CNY tr-TR": { code: "CNY", symbol: "CN¥", decimalPlaces: 2, name: "Çin yuanı" },
    };

    const result = {
      CNY: getCurrency("CNY", "en-US"),
      "CNY zh-CN": getCurrency("CNY", "zh-CN"),
      "CNY zh-TW": getCurrency("CNY", "zh-TW"),
      "CNY tr-TR": getCurrency("CNY", "tr-TR"),
    };

    expect(result).toEqual(expected);
  });
});


rollup.config.ts
/* eslint-disable import/no-unresolved, global-require, @typescript-eslint/no-var-requires , import/no-extraneous-dependencies */
import resolve from "rollup-plugin-node-resolve";
import commonjs from "rollup-plugin-commonjs";
import sourceMaps from "rollup-plugin-sourcemaps";
import camelCase from "lodash.camelcase";
import typescriptPlugin from "rollup-plugin-typescript2";
import json from "rollup-plugin-json";
import { terser } from "rollup-plugin-terser";
import { parse } from "path";
import bundleSize from "rollup-plugin-bundle-size";
import typescript from "typescript";

const pkg = require("./package.json");

const libraryName = parse(pkg.main).name;

export default {
  input: `src/${libraryName}.ts`,
  output: [
    { file: pkg.main, name: camelCase(libraryName), format: "cjs", sourcemap: true },
    { file: pkg.main.replace(".js", ".min.js"), name: camelCase(libraryName), format: "cjs", sourcemap: true },
    { file: pkg.module, format: "esm", sourcemap: true },
    { file: pkg.module.replace(".mjs", ".min.mjs"), format: "esm", sourcemap: true },
    { file: pkg["umd:main"], name: camelCase(libraryName), format: "umd", sourcemap: true },
    { file: pkg["umd:main"].replace(".js", ".min.js"), name: camelCase(libraryName), format: "umd", sourcemap: true },
  ],
  // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
  external: [...Object.keys(pkg.dependencies || {})],
  watch: {
    include: "src/**",
  },
  plugins: [
    bundleSize(),
    terser({
      include: [/^.+\.min\.m?js$/],
      exclude: [],
    }),
    // Allow json resolution
    json(),
    // Compile TypeScript files
    typescriptPlugin({ useTsconfigDeclarationDir: true, typescript }),
    // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
    commonjs(),
    // Allow node_modules resolution, so you can use 'external' to control
    // which external modules to include in the bundle
    // https://github.com/rollup/rollup-plugin-node-resolve#usage
    resolve(),

    // Resolve source maps to the original source
    sourceMaps(),
  ],
};


src/intl-data.ts
/** ignore */
type LocaleCode = string;
/** ignore */
type CurrencyCode = string;

/** Information about locale. */
export interface Locale {
  /** BCP 47 language tag of the locale. This is simply string provided to get locale. Example: `tr-TR` */
  code: LocaleCode;
  /** Character to separate decimal (fraction) parts of the number from integer parts. */
  decimalCharacter: string;
  /** Grouping separators for integer parts, such as thousands separators or thousand/lakh/crore separators. */
  digitGroupSeparator: string;
  /** Place of currency symbol: `p` for prefix such as `$1.95`, or `s` for suffix such as `1.95$` */
  currencySymbolPlacement: "p" | "s";
  /** Place of percent symbol: `p` for prefix such as `%12`, or `s` for suffix such as `12%` */
  percentSymbolPlacement: "p" | "s";
  /** Percent symbol. Some locales contains spcae for percent symbol such as `12 %`, and some not `12%`. This is used to differenciate them. */
  percentSymbol: string;
}

/** Information about currency */
export interface Currency {
  /** Name of the currency. */
  name: string;
  /**  ISO 4217 currency codes */
  code: CurrencyCode;
  /** Number of digits used in decimal (fractional) part of the currency. */
  decimalPlaces: number;
  /** Currency symbol such as `₺`, `$`, `€` */
  symbol: string;
}

const localeCache: Record<LocaleCode, Locale> = {};
const currencyCache: Record<CurrencyCode, Record<LocaleCode, Currency>> = {};

export function getCurrency<T extends null | undefined>(currencyCode: T, locale?: string): T;
export function getCurrency(currencyCode: string | null | undefined, locale?: string): Currency | null | undefined;
/**
 * Returns currency details for given currency and optionally locale. Throws error for unknown currency codes.
 *
 * @param currencyCode is the currency to use in currency formatting. Possible values are the ISO 4217 currency codes, such as "USD" for the US dollar.
 * @param locale is the optional string with a BCP 47 language tag. If provided returned object contains currency name in given locale.
 * @returns currency details for requested currency. Also returns `null` for `null` input and `undefined` for `undefined` input.
 */
export function getCurrency(
  currencyCode: string | null | undefined,
  locale = new Intl.NumberFormat().resolvedOptions().locale
): Currency | null | undefined {
  if (typeof currencyCode !== "string") {
    return currencyCode === null ? null : undefined;
  }

  const commonOptions = { style: "currency", currency: currencyCode };
  currencyCache[currencyCode] = currencyCache[currencyCode] || {};

  // Compute currency data
  if (currencyCache[currencyCode][locale] === undefined) {
    const textWithNumbers = (0).toLocaleString(locale, { ...commonOptions, currencyDisplay: "symbol" }).replace(/[.,]/, ""); // "OMR 0000" or "$000" or "RWF 0" or "0000 OMR"... without decimal separator
    // eslint-disable-next-line no-irregular-whitespace
    if ((0).toLocaleString(locale, { ...commonOptions, currencyDisplay: "name" }).includes(currencyCode)) {
      throw new Error(`Invalid currency code: ${currencyCode}.`);
    }
    const lengthWithNumbers = textWithNumbers.length;
    const textWithoutNumbers = textWithNumbers.replace(/0/g, "");
    const decimalPlaces = lengthWithNumbers - textWithoutNumbers.length - 1;
    const symbol = textWithoutNumbers.trim();
    const formattedCurrency = (0).toLocaleString(locale, { ...commonOptions, currencyDisplay: "name" });
    const formattedNumber = (0).toLocaleString(locale, { minimumFractionDigits: decimalPlaces });
    const rxSymbol = new RegExp(` ?${symbol} ?`);
    const rxNumber = new RegExp(` ?${formattedNumber} ?`);
    const name = formattedCurrency.replace(rxSymbol, "").replace(rxNumber, "");
    currencyCache[currencyCode][locale] = { name, code: currencyCode, symbol, decimalPlaces };
  }

  return currencyCache[currencyCode][locale];
}

export function getLocale(localeCode: string): Locale;
export function getLocale<T extends null | undefined>(localeCode: T): T;
export function getLocale(localeCode: string | null | undefined): Locale | undefined | null;
/**
 * Returns locale details for given locale. Throws error for unsupported locales.
 *
 * @param localeCode is the string with a BCP 47 language tag. If provided returned object contains currency name in given locale.
 * @returns locale data for requested locale. Also returns `null` for `null` input and `undefined` for `undefined` input.
 */
export function getLocale(localeCode: string | null | undefined): Locale | undefined | null {
  if (typeof localeCode !== "string") {
    return localeCode === null ? null : undefined;
  }

  if (Intl.NumberFormat.supportedLocalesOf(localeCode, { localeMatcher: "lookup" }).length === 0) {
    throw new Error(`Unsupported locale: ${localeCode}.`);
  }

  if (localeCache[localeCode] === undefined) {
    const [digitGroupSeparator, decimalCharacter] = (11111.1).toLocaleString(localeCode).replace(/1/g, "");
    const currencySymbolPlacement = (1).toLocaleString(localeCode, { style: "currency", currency: "USD" })[0] === "1" ? "s" : "p";
    const percentString = (0.01).toLocaleString(localeCode, { style: "percent" });
    const percentSymbolPlacement = percentString[0] === "1" ? "s" : "p";
    const percentSymbol = percentString.replace("1", ""); // Some locales contains space, some not.
    localeCache[localeCode] = {
      code: localeCode,
      digitGroupSeparator,
      decimalCharacter,
      currencySymbolPlacement,
      percentSymbolPlacement,
      percentSymbol,
    };
  }
  return localeCache[localeCode];
}
